linkerd/app/core/src/dns.rs

```diff
@@ -8,16 +17,93 @@ pub struct Config {
 }
 
 pub struct Dns {
-    pub resolver: Resolver,
+    resolver: Resolver,
+    dns_resolutions_total: Family<Labels, Counter>,
+}
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+struct Labels {
+    client: &'static str,
+    record_type: RecordType,
+    result: Outcome,
+}
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+enum RecordType {
+    A,
+    Srv,
+}
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+enum Outcome {
+    Ok,
+    NotFound,
+}
+
+// === impl Dns ===
+
+impl Dns {
+    /// Returns a new [`Resolver`].
+    pub fn resolver(&self, client: &'static str) -> Resolver {
+        let metrics = self.metrics(client);
+
+        self.resolver.clone().with_metrics(metrics)
+    }
+
+    fn metrics(&self, client: &'static str) -> Metrics {
+        let family = &self.dns_resolutions_total;
+
+        let a_records_resolved = (*family.get_or_create(&Labels {
+            client,
+            record_type: RecordType::A,
+            result: Outcome::Ok,
+        }))
+        .clone();
+        let a_records_not_found = (*family.get_or_create(&Labels {
+            client,
+            record_type: RecordType::A,
+            result: Outcome::NotFound,
+        }))
+        .clone();
+        let srv_records_resolved = (*family.get_or_create(&Labels {
+            client,
+            record_type: RecordType::Srv,
+            result: Outcome::Ok,
+        }))
+        .clone();
+        let srv_records_not_found = (*family.get_or_create(&Labels {
+            client,
+            record_type: RecordType::Srv,
+            result: Outcome::NotFound,
+        }))
+        .clone();
+
+        Metrics {
+            a_records_resolved,
+            a_records_not_found,
+            srv_records_resolved,
+            srv_records_not_found,
+        }
+    }
 }
 
 // === impl Config ===
 
 impl Config {
-    pub fn build(self) -> Dns {
+    pub fn build(self, registry: &mut Registry) -> Dns {
+        let dns_resolutions_total = Family::default();
+        registry.register(
+            "dns_resolutions_total",
+            "Counts the number of DNS records that have been resolved.",
+            dns_resolutions_total.clone(),
+        );
```

Comment:
I'm fairly sure this will result in a metric called `dns_resolutions_total_total`: counters implicitly have _total appended (but the metric name in the HELP does not include the _total suffix).

Furthermore, as we do elsewhere, we should omit common prefixes like dns_ and set them from the caller.

```suggestion
        registry.register(
            "resolutions",
            "Counts the number of DNS records that have been resolved.",
            dns_resolutions_total.clone(),
        );
```

I'd also probably opt to name the field `resolutions` but that's a pure nit.

Reply:
1437755 addresses this; removing the `_total` suffix and omitting the `dns_` prefix locally in favor of a sub-registry.